Library/Homebrew/cmd/shellutils

```diff
@@ -0,0 +1,303 @@
+        a) whence_next=continue ;;
+        h) whence_usage; exit 0 ;;
+        s) whence_silent=1      ;;
+        v) whence_verbose=1     ;;
+       \?) whence_usage; exit 1 ;;
+      esac
+  done; shift $((OPTIND - 1))
+
+  whence_try()
+  {
+    echo "$PATH:." | tr ':' '\n' | while read whence_component; do
+
+      whence_try="$whence_component/$whence_util"
+      [ -e "$whence_try" ] && [ -x "$whence_try" ] &&
+        echo "$whence_try" && $whence_next || continue
+    done
+  }
+
+  for whence_util; do :; done
+  [ -"$whence_util"- = -""- ] && whence_usage && exit 1
+
+  whence_result="$(whence_try "$whence_util")"
+
+  [ -"$whence_result"- = -""- ] && whence_error && exit 1 ||
+    [ -$whence_silent- = -1- ] && exit 0 || echo "$whence_result" | uniq
+}
+
+which(){ whence "$@" ;}
+
+##===--------------------------------------------------------------------===##
+
+# Override TERM if not outputting to a TTY.
+
+brew_shellutils_term_init()
+{
+  [ -t 1 ] && [ -t 2 ] || { TERM=dumb; export TERM ;}
+}
+
+#===--------------------------------------------------------------------===##
+
+brew_shellutils_help(){
+
+  echo "
+  $(bold)Available in $(underline)brew shellutils$(cclear) (source this file):
+
+  - $(bold)$(blue)realpath$(cclear)
+  - $(bold)$(blue)readlink$(cclear) [-efmnqsvz]
+
+    $(bold)Fully dereference symbolic links. Compatible with GNU readlink.
+    ($(cclear)brew shellutils is $(underline)$(realpath "$0")$(cclear)"\)."
+    $(cclear)$(bold)
+  - $(blue)whence$(cclear)$(bold)
+  - $(blue)which $(cclear)[-asv]
+
+    $(bold)Like /usr/bin/which. $(underline)Not a ksh \`whence\`$(cclear).
+    $(cclear)$(bold)
+  - $(blue)ohai$(cclear)$(bold)
+  - $(blue)note$(cclear)$(bold)
+       " >&2
+  note "Pretty-print a message to standard error."
+
+   echo "$(bold)
+  - $(blue)opoo$(cclear)$(bold)
+  - $(blue)warn$(cclear)$(bold)
+        " >&2
+
+  warn "$(yellow)Pretty-print a warning to standard error."
+
+  echo "$(bold)
+  - $(blue)odie$(cclear)$(bold)
+  - $(blue)die$(cclear)$(bold)
+       " >&2
+
+  echo " $(red)⨷  $(bold)Pretty-print an error to standard error, exit 1." >&2
+
+  echo "$(cclear)
+  - $(bold)$(red)red$(cclear)
+  - $(bold)$(yellow)yellow$(cclear)
+  - $(bold)$(magenta)magenta$(cclear)
+  - $(bold)$(cyan)cyan$(cclear)
+  - $(bold)$(green)green$(cclear)
+
+  $(bold) These use tput for $(underline)safe$(cclear)$(bold) terminal colors.
+  $(cclear)" >&2
+ cclear
+}
+
+##===--------------------------------------------------------------------===##
+
+# Intitialization and help.
+
+brew_shellutils_main(){
+ brew_shellutils_term_init
+ brew_shellutils_help "$@"
+}
+
+##===--------------------------------------------------------------------===##
+##===--- end helper functions -------------------------------------------===##
+##===--------------------------------------------------------------------===##
+
+case $(basename "$0") in
+   *shellutils) brew_shellutils_main "$@" ;;
+   *|'') continue ;;
+esac
```

Comment:
I'd rather we just tried to do it in raw bash and worry about making it look pretty at a later pass.